demo/lego_official_projects/common.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python
+
+import logging
+import sys
+import time
+from ev3dev.auto import *
+
+log = logging.getLogger(__name__)
+
+
+class Tank(object):
+
+    def __init__(self, left_motor, right_motor):
+
+        for motor in (left_motor, right_motor):
+            if motor not in OUTPUTS:
+                log.error("%s in an invalid output, choices are %s" % (motor, ''.join(OUTPUTS)))
+                sys.exit(1)
+
+        self.left_motor = LargeMotor(left_motor)
+        self.right_motor = LargeMotor(right_motor)
+
+        for x in (self.left_motor, self.right_motor):
+            if not x.connected:
+                log.error("%s is not connected" % x)
+                sys.exit(1)
+
+        self.left_motor.reset()
+        self.right_motor.reset()
+        self.speed = 90
+        self.left_motor.duty_cycle_sp = self.speed
+        self.right_motor.duty_cycle_sp = self.speed
+
+
+class RemoteControlledTank(Tank):
+
+    def __init__(self, left_motor, right_motor):
+        Tank.__init__(self, left_motor, right_motor)
+        self.remote = RemoteControl(channel=1)
+
+        if not self.remote.connected:
+            log.error("The remote/IR-Sensor is not connected")
+            sys.exit(1)
+
+        self.remote.on_red_up = self.move_left_track_forward
+        self.remote.on_red_down = self.move_left_track_backward
+        self.remote.on_blue_up = self.move_right_track_forward
+        self.remote.on_blue_down = self.move_right_track_backward
+
+    def move_left_track_forward(self, state):
+        if state:
+            self.left_motor.run_forever(duty_cycle_sp=self.speed)
+        else:
+            self.left_motor.stop()
+
+    def move_right_track_forward(self, state):
+        if state:
+            self.right_motor.run_forever(duty_cycle_sp=self.speed)
+        else:
+            self.right_motor.stop()
+
+    def move_left_track_backward(self, state):
+        if state:
+            self.left_motor.run_forever(duty_cycle_sp=self.speed * -1)
+        else:
+            self.left_motor.stop()
+
+    def move_right_track_backward(self, state):
```

Comment:
That is not exactly what I had in mind. The functions look very similar and only differ by the motor they are applied to and the direction. So all four of those could be replaced by
```py
def make_move(motor, dc_sp):
    def move(state):
        if state:
            motor.run_forever(duty_cycle_sp=dc_sp)
        else:
            motor.stop()
    return move
```

Then,
```py
    self.remote.on_red_up = self.make_move(self.left_motor, self.speed)
    ...
```